File: frontend/lib/core/models/book.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'book.freezed.dart';
part 'book.g.dart';

/// Конвертер для преобразования ID из int (PostgreSQL) или String (UUID) в String
String _bookIdToString(dynamic value) {
  return value.toString();
}

/// Конвертер для преобразования child_id из int (PostgreSQL) или String (UUID) в String
String _childIdToString(dynamic value) {
  if (value == null) {
    throw ArgumentError('child_id не может быть null');
  }
  if (value is int) {
    return value.toString();
  }
  if (value is String) {
    return value;
  }
  return value.toString();
}

/// Конвертер для безопасного парсинга даты из JSON
DateTime _dateTimeFromJson(dynamic value) {
  if (value == null) {
    throw ArgumentError('created_at не может быть null');
  }
  if (value is DateTime) {
    return value;
  }
  if (value is String) {
    try {
      return DateTime.parse(value);
    } catch (e) {
      print('[Book] Ошибка парсинга даты: $value, ошибка: $e');
      // Если не удалось распарсить, возвращаем текущую дату
      return DateTime.now();
    }
  }
  if (value is int) {
    // Unix timestamp в секундах
    return DateTime.fromMillisecondsSinceEpoch(value * 1000);
  }
  print('[Book] Неожиданный тип для created_at: ${value.runtimeType}, значение: $value');
  return DateTime.now();
}

@freezed
class Book with _$Book {
  const factory Book({
    // обязательные поля
    @JsonKey(fromJson: _bookIdToString) required String id,
    @JsonKey(name: 'child_id', fromJson: _childIdToString) required String childId,
    @JsonKey(name: 'user_id') String? userId,
    required String title,

    @JsonKey(name: 'created_at', fromJson: _dateTimeFromJson)
    required DateTime createdAt,

    // статус книги (draft / editing / final)
    @JsonKey(name: 'status')
    @Default('draft')
    String status,

    // необязательные поля
    @JsonKey(name: 'cover_url') String? coverUrl,
    @JsonKey(name: 'final_pdf_url') String? finalPdfUrl,
    
    // Статус оплаты книги
    @JsonKey(name: 'is_paid') @Default(false) bool isPaid,

    @JsonKey(name: 'pages')
    List<dynamic>? pages,

    @JsonKey(name: 'edit_history')
    List<dynamic>? editHistory,

    @JsonKey(name: 'images_final')
    List<dynamic>? imagesFinal,
  }) = _Book;

  factory Book.fromJson(Map<String, dynamic> json) => _$BookFromJson(json);
}

File: backend/app/models/book.py
from sqlalchemy import Column, String, Integer, Text, DateTime, ForeignKey, JSON
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from ..db import Base


class Book(Base):
    __tablename__ = "books"

    id = Column(UUID(as_uuid=True), primary_key=True, index=True, server_default=func.gen_random_uuid())
    child_id = Column(Integer, ForeignKey("children.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(String, nullable=True)
    
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    genre = Column(String, nullable=True)
    theme = Column(String, nullable=True)
    writing_style = Column(String, nullable=True)
    narrator = Column(String, nullable=True)
    
    cover_url = Column(Text, nullable=True)
    content = Column(Text, nullable=True)
    pages = Column(JSONB, nullable=True)
    prompt = Column(Text, nullable=True)
    ai_model = Column(String, nullable=True)
    variables_used = Column(JSONB, nullable=True)
    audio_url = Column(Text, nullable=True)
    
    status = Column(String, nullable=False, default="draft")
    
    # Workflow fields
    final_pdf_url = Column(Text, nullable=True)
    images_final = Column(JSONB, nullable=True)
    edit_history = Column(JSONB, nullable=True)
    detail_prompt = Column(Text, nullable=True)
    is_paid = Column(String, nullable=True, default="false")  # "true" или "false" как строка для совместимости
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    scenes = relationship("Scene", back_populates="book", cascade="all, delete-orphan")
